### DISCUSSION TITLE
SSD1306 display doesn't work

### DISCUSSION THREAD

**Discussion by undefinedCat** [micropython/micropython#10605] [ESP8266]
I have a 128x32 display, and this is what I typed into the REPL:
```
>>> from machine import Pin, I2C
>>> import ssd1306
>>> i2c = I2C(scl=Pin(5), sda=Pin(4), freq=100000)
>>> screen = ssd1306.SSD1306_I2C(128, 32, i2c)
```
And after that it shows this error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "ssd1306.py", line 118, in __init__
  File "ssd1306.py", line 37, in __init__
  File "ssd1306.py", line 74, in init_display
  File "ssd1306.py", line 123, in write_cmd
```
What's the solution to this problem?

**Comment by robert-hh** [comment]
It looks like a basic communication problem. Call `print(i2c.scan())` after instantiating i2c to see, if the display is detected and at which address.

**Reply by NeuralShan** [reply]
I have been facing the same issue, and tried calling the .scan() function, but it returned a blank array [ ].

**Reply by peterhinch** [reply]
If `.scan()` returns `[]` there is a hardware problem. Incorrect connections, lack of pullup resistors or a device problem.

**Comment by karfas** [comment]
> What's the solution to this problem?

This is hard to tell without seeing the error message you got.

However, I would assume you have some kind of connection problem.
Does `i2c.scan()` find the display at the correct address ?
Do you have proper pull-up resistors in place ?

**Comment by shariltumin** [comment]
I have an i2c ssd1306 oled display and can confirm that I have the same problem.

The display works on esp32, esp32s2 but not esp8266.

I have made a copy of "ssd1306.py" to "oled.py" so that I can put in some debugging messages.

The ssd1306 module reports ENODEV during init:
```py
MicroPython v1.22.0-preview.246.g4365edb81.dirty on 2024-01-15; WEMOS D1 MINI 4MB with ESP8266
>>> 
>>> from machine import Pin, SoftI2C
>>> i2c = SoftI2C(sda=Pin(4), scl=Pin(5))
>>> i2c.scan()
[60]
>>> import oled
>>> display = oled.SSD1306_I2C(128, 64, i2c)
Write CMD: 60 bytearray(b'\x80\xae')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "oled.py", line 119, in __init__
  File "oled.py", line 38, in __init__
  File "oled.py", line 75, in init_display
  File "oled.py", line 125, in write_cmd
OSError: [Errno 19] ENODEV
>>> hex(60)
'0x3c'
>>> i2c.scan()
[60]
```

**Reply by shariltumin** [reply]
Sorry, please disregard my post above. It seems that the board I did the test on was defective. Doing the test on another esp8266 boards worked fine.

**Reply by BryanSouza91** [reply]
> Sorry, please disregard my post above. It seems that the board I did the test on was defective. Doing the test on another esp8266 boards worked fine.

I would verify that the SDA and SCL pins are not reversed on the "defective" board. Refer to my comment below for my experience.

**Comment by zuofeng64** [comment]
screen = ssd1306.SSD1306_I2C(128, 32, i2c)
这句是否错了，
screen = ssd1306.SSD1306_I2C(128, 64, i2c)

**Comment by BryanSouza91** [comment]
I experienced this same error with the esp8266 board with builtin ssd1306 display that I have. The I2C pins are labeled on the underside of the board as `D1 = SDA D2 = SCL` this means `D1 = Pin(5) D2 = Pin(4)` which is backwards from the ssd1306 defaults in the doc examples. 

`i2c.scan()` did **not** work when initialized as the example is

the display works fine for me when initialized as `i2c = I2C(sda=Pin(5), scl=Pin(4))`

**Comment by bidrohini** [comment]
The I2C address specified in the ssd1306 library may not match the address of your display. Run the I2C scanner to check the connected devices and their addresses.If the I2C scanner shows an address different from 0x3C, update the 'addr' parameter. You can also check the display part of this tutorial for further cross-checking. https://www.pcbway.com/project/shareproject/Raspberry_Pico_with_Ultrasonic_Sensor_and_SSD1306.html

**Comment by be80be** [comment]
The problem I found was the ESP8266 I was using had the power pins miss labelled 
Its a wiring problem I bet
